ASMDocumentation.md

```diff
@@ -0,0 +1,5841 @@
++        mov     eax, 1
++        test    eax, eax
++        je      SHORT $LN3@LockClass
++        push    0
++        mov     ecx, DWORD PTR _nFlag$[ebp]
++        call    unsigned int std::_Atomic_storage<unsigned int,4>::load(std::memory_order)const  ; std::_Atomic_storage<unsigned int,4>::load
++        mov     DWORD PTR _expected$1[ebp], eax
++        push    2
++        mov     ecx, DWORD PTR _expected$1[ebp]
++        or      ecx, 1
++        push    ecx
++        lea     edx, DWORD PTR _expected$1[ebp]
++        push    edx
++        mov     ecx, DWORD PTR _nFlag$[ebp]
++        call    bool std::atomic<unsigned int>::compare_exchange_weak(unsigned int &,unsigned int,std::memory_order) ; std::atomic<unsigned int>::compare_exchange_weak
++        movzx   eax, al
++        test    eax, eax
++        je      SHORT $LN4@LockClass
++        jmp     SHORT $LN3@LockClass
++        jmp     SHORT $LN6@LockClass
++$LN4@LockClass:
++        mov     ecx, DWORD PTR _expected$1[ebp]
++        and     ecx, 1
++        je      SHORT $LN6@LockClass
+         call    static void ThreadClass::Switch_Thread(void)       ; ThreadClass::Switch_Thread
+-        mov     ebx, DWORD PTR _nFlag$[ebp]
+-        DB      -16                       ; fffffff0H
+-        bts     DWORD PTR [ebx], 0
+-        jb      SHORT $The_Bit_Was_Previously_Set_So_Try_Again$5
+-$BitSet$4:
++        npad    1
++$LN6@LockClass:
++        jmp     SHORT $LN8@LockClass
++$LN3@LockClass:
+         mov     eax, DWORD PTR _this$[ebp]
+-        pop     edi
+-        pop     esi
+-        pop     ebx
+         mov     esp, ebp
+         pop     ebp
+         ret     4
+```
+
+</details>
+
+---
+
+<details>
+<summary>Generals/Code/Libraries/Source/WWVegas/WWDebug/wwdebug.h</summary>
+
+This header defines the following with assembly:
+
+```c++
+#ifdef WWDEBUG
+#define WWDEBUG_BREAK							_asm int 0x03
+#else
+#define WWDEBUG_BREAK							_asm int 0x03
+#endif
+```
+
+This is redundant as it is, but `_asm int 0x03` is equivalent to `__debugbreak();` on Win32:
+
+```c++
+#include <intrin.h>
+
+#define WWDEBUG_BREAK __debugbreak();
+```
+
+> **NOTE**: Observe that for this program:
+>
+> ```c++
+> #include <intrin.h>
+> 
+> #define WWDEBUG_BREAK							_asm int 0x03
+> 
+> void first() {
+>     WWDEBUG_BREAK;
+> }
+> 
+> void second() {
+>     __debugbreak();
+> }
+> ```
+>
+> The resulting assembly is:
+>
+> ```asm
+> void first(void) PROC                                  ; first
+>         push    ebp
+>         mov     ebp, esp
+>         int     3
+>         pop     ebp
+>         ret     0
+> void first(void) ENDP                                  ; first
+> 
+> void second(void) PROC                           ; second
+>         push    ebp
+>         mov     ebp, esp
+>         int     3
+>         pop     ebp
+>         ret     0
+> void second(void) ENDP                           ; second
+> ```
+>
+> These `#ifdef` blocks will be useful for cross-platform compatibility:
+>
+> ```c++
```

Comment:
[here](https://github.com/TheAssemblyArmada/Thyme/blob/ccef1e11c1355c6db577a057c06e7d790f1a0333/deps/baseconfig/src/intrinsics.h#L141) Thyme implements this for several CPUs.